shim/src/epoll/epoll.c

```diff
@@ -42,3 +53,61 @@ struct demi_event *epoll_get_event(int epfd, int i)
     assert(epoll_table[epfd].used == 1);
     return (&epoll_table[epfd].events[i]);
 }
+
+struct demi_event *epoll_get_head(int epfd)
+{
+    assert(epoll_table[epfd].used == 1);
+
+    if (epoll_table[epfd].head == INVALID_EV)
+        return NULL;
+
+    return (&epoll_table[epfd].events[epoll_table[epfd].head]);
+}
+
+struct demi_event *epoll_get_tail(int epfd)
+{
+    assert(epoll_table[epfd].used == 1);
+
+    if (epoll_table[epfd].tail == INVALID_EV)
+        return NULL;
+
+    return (&epoll_table[epfd].events[epoll_table[epfd].tail]);
+}
+
+void epoll_set_head(int epfd, int i)
+{
+    assert(epoll_table[epfd].used == 1);
+
+    epoll_table[epfd].head = i;
+}
+
+void epoll_set_tail(int epfd, int i)
+{
+    assert(epoll_table[epfd].used == 1);
+
+    epoll_table[epfd].tail = i;
+}
+
+struct demi_event *epoll_get_next(int epfd, struct demi_event *ev)
+{
+    int i;
+
+    if (ev == NULL || ev->next_ev == INVALID_EV)
+        return NULL;
+
+    i = ev->next_ev;
+
+    return (epoll_get_event(epfd, i));
+}
+
+struct demi_event *epoll_get_prev(int epfd, struct demi_event *ev)
+{
+    int i;
+
+    if (ev == NULL || ev->prev_ev == INVALID_EV)
+        return NULL;
+
+    i = ev->prev_ev;
+
+    return (epoll_get_event(epfd, i));
```

Comment:
Inline this variable. This is not for performance reason, but to reduce surface of bugs :)